storage/SMAIndexSubBlock.hpp

```diff
@@ -0,0 +1,432 @@
+   *
+   * @param attribute The ID of the attribute to check.
+   * 
+   * @return \c true if this index contains an entry.
+   */
+  bool hasEntryForAttribute(attribute_id attribute) const;
+
+  /**
+   * @brief Returns the number of tuples, the aggregate COUNT, of the storage sub block.
+   *
+   * @return Number of tuples in the sub block.
+   */
+  std::uint32_t getCount() const {
+    return reinterpret_cast<sma_internal::SMAHeader*>(sub_block_memory_)->count_aggregate;
+  }
+
+ private:
+  inline sma_internal::Selectivity selectivityForPredicate(const ComparisonPredicate &predicate) const;
+
+  // Retrieves an entry, first checking if the given attribute is indexed.
+  inline const sma_internal::SMAEntry* getEntryChecked(attribute_id attribute) const {
+    DCHECK(attribute > -1) << "Attribute Id must be >= 0";
+
+    if (total_attributes_ > static_cast<std::size_t>(attribute)) {
+      int index = attribute_to_entry_[attribute];
+      if (index != -1) {
+        return entries_ + index;
+      }
+    }
+    return nullptr;
+  }
+
+  // Retrieves an entry, not checking if the given attribute is indexed.
+  // Warning: This should not be used unless the attribute is indexed.
+  inline const sma_internal::SMAEntry* getEntry(attribute_id attribute) const {
+    return entries_ + attribute_to_entry_[attribute];
+  }
+
+  // Resets a single entry to a zero and invalid state.
+  // This should be called before rebuilding, or while initializing.
+  void resetEntry(sma_internal::SMAEntry *entry, attribute_id attribute, const Type &attribute_type);
+
+  // Sets all entries to a zero'd and invalid state.
+  // This is called prior to a rebuild.
+  void resetEntries();
+
+  void addTuple(tuple_id tuple);
+
+  // Frees any TypedValue data which is held in the SMA entries.
+  void freeOutOfLineData();
+
+  // Handy pointer to the header.
+  sma_internal::SMAHeader *header_;
+  // Handy pointer to the beginning of the entries.
+  sma_internal::SMAEntry *entries_;
+  // Total number of attributes in the relation.
+  std::size_t total_attributes_;
+  // Used to lookup the index of the SMA entry given the attribute_id.
+  // For example SMAEntry &entry =
+  //                 entries_[attribute_to_entry_[someAttribute->getID()]]
+  // A value of -1 indicates that that attribute is not indexed.
+  int *attribute_to_entry_;
+  // Number of indexed attributes.
+  std::size_t indexed_attributes_;
+  // True if the index has gone through the initialization process.
+  bool initialized_;
+
+  // An array of pointers to necessary add operations for updating SUM.
+  // Essentially, it maps between TypeID and the pointer to the correct
+  // operator for that type.
+  //    For example: add_operations_[AttributeTypeID]->applyWithTypedValues(
+  //                                          TypedValue of the attributes type,
+  //                                          TypedValue of the sum type)
+  UncheckedBinaryOperator** add_operations_;
```

Comment:
I would prefer an unordered_map with key as ``AttributeTypeID`` and value as a ``unique_ptr<UncheckedBinaryOperator>``. Makes it less cumbersome to manage the memory and easy to understand. 

If the IDs follow a sequence, you can use a PtrVector of UncheckedBinaryOperators too.